internal/config/config.go

```diff
@@ -0,0 +1,163 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package config
+
+import (
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+
+	"sigs.k8s.io/kubebuilder/pkg/model/config"
+	"sigs.k8s.io/yaml"
+)
+
+// Default path for the configuration file
+const DefaultPath = "PROJECT"
+
+func exists(path string) (bool, error) {
+	// Look up the file
+	_, err := os.Stat(path)
+
+	// If we could find it the file exists
+	if err == nil {
+		return true, nil
+	}
+
+	// Not existing and different errors are differentiated
+	if os.IsNotExist(err) {
+		err = nil
+	}
+	return false, err
+}
+
+// Exists verifies that the configuration file exists in the default path
+// TODO: consider removing this verification in favor of using Load and checking the error
+func Exists() (bool, error) {
+	return exists(DefaultPath)
+}
+
+func readFrom(path string) (c config.Config, err error) {
+	// Read the file
+	in, err := ioutil.ReadFile(path) // nolint: gosec
+	if err != nil {
+		return
+	}
+
+	// Unmarshal the file content
+	if err = yaml.Unmarshal(in, &c); err != nil {
+		return
+	}
+
+	// kubebuilder v1 omitted version, so default to v1
+	if c.Version == "" {
+		c.Version = config.Version1
+	}
+
+	return
+}
+
+// Read obtains the configuration from the default path but doesn't allow to persist changes
+func Read() (*config.Config, error) {
+	return ReadFrom(DefaultPath)
+}
+
+// ReadFrom obtains the configuration from the provided path but doesn't allow to persist changes
+func ReadFrom(path string) (*config.Config, error) {
+	c, err := readFrom(path)
+
+	return &c, err
+}
+
+// Config extends model/config.Config allowing to persist changes
+// NOTE: the existence of Config structs in both model and internal packages is to guarantee that kubebuilder
+// is the only project that can modify the file, while plugins can still receive the configuration
```

Comment:
My opinion here is that kubebuilder should provide some hooks to plugins, so that the config file access is still centralized. Advantages: handle authorization, avoid races that may end in unparsable config files, abstract plugins from the config persistance implementation (who knows if we may want to move from a YAML file to a JSON file or a DB in the future).

Reply:
Do you have an ad-hoc example of what a hook would look like?

Authorization would be handled by the `Read`/`Load` semantic differences; data races can be handled by mutexes; type of persistence doesn't matter because the underlying config is loaded into a Go type, which won't change much.

Reply:
Authorithation: that is just a boolean, either be able to edit or not, will may be problematic. We may want finer control.
Data races: plugins were planned to support separate processes, weren't them?
Hooks: didn't think too much about this. In-process plugins could have access to a `chan Config` that they can send back to kubebuilder with the changes for example.